nodes/sensing_node.py

```diff
@@ -55,48 +64,70 @@ def __init__(self):
         )
 
     def _callback(self, message):
-        print("Enter callback function of Image Handler")
+        # print("Enter callback function of Image Handler")
         try:
             cv_image = self.bridge.imgmsg_to_cv2(message, "mono8")
         except CvBridgeError as e:
             rospy.logerr("Error in imageCallback: %s", e)
 
-        # detect lines
+        # Detect lines
         cv_image_color = cv2.cvtColor(cv_image, cv2.COLOR_GRAY2BGR)
-        cv2.imwrite("cv_im_color.jpg", cv_image_color)
         #  Canny-Edge Detector
         canny_image = cv2.Canny(cv_image_color, 110, 200)
-        cv2.imwrite("canny_image.jpg", canny_image)
+        self.latest_canny_image = canny_image
+
+        # generate color image and draw box on road
+        roi_left_line_im = self.image_helper.road2image(self.roi_left_line)
+        roi_right_line_im = self.image_helper.road2image(self.roi_right_line)
+        cv2.polylines(
+            cv_image_color,
+            [roi_left_line_im.astype(np.int32), roi_right_line_im.astype(np.int32)],
+            isClosed=True,
+            color=(0, 0, 255),
+            thickness=8,
+        )
+        # downscale to reduce load
+        cv_image_color = cv2.pyrDown(cv_image_color)
+        try:
+            self.pub_dbg_image.publish(
+                self.bridge.cv2_to_imgmsg(cv_image_color, "bgr8")
+            )
+        except CvBridgeError as e:
+            rospy.logerr(e)
+
+        # print("Exit callback function of Image Handler")
+
+    def estimate_lane(self, Z_initial):
         #  Convert detected edges, that are in image coordinates, to road coordinates.
+        height_image = self.latest_canny_image.shape[0]
         pts_im = np.array([])
-        for x, y in np.ndindex(canny_image.shape):
-            if canny_image[x, y] == 255:
-                pts_im = np.append(pts_im, [y, x])
+        crop_idx = np.array([height_image / 2, height_image * 0.74]).astype(int)
+        canny_cropped = np.array(
+            self.latest_canny_image[crop_idx[0] : crop_idx[1], :]  # noqa:E203
+        )
+        for x in range(0, len(canny_cropped), 5):
+            x += crop_idx[0]
+            for y in range(0, len(canny_cropped[0]), 2):
+                if self.latest_canny_image[x, y] == 255:
+                    pts_im = np.append(pts_im, [y, x])
```

Comment:
This is "ok" as is for this PR, but we may be able to improve performance by using `np.argwhere()`. In essence what we do here is, "give me the x,y coords of every point above the threshold of 255". As most numpy functions are implemented in machine code, we might see considerable performance gains.

@sunny-5555 Try it out and put in the next PR, I'll accept it as is for now.

Reply:
Like this:
```python
np.argwhere(self.latest_canny_image > 255)
```